rotary_utils/repair.py

```diff
@@ -55,59 +54,222 @@ def main(args):
+        """
+        """
+        Description of attributes:
+            linking_outdir_base: Temporary directory where re-assembly and stitching files will be saved.
+            circular_contig_tmp_fasta: Temporary FastA file containing only circular contigs to be stitched.
+            bam_filepath: path to a BAM file with mapping information of long reads to the contigs (it is OK if this
+                          file also contains mappings to other contigs outside those in the input file).
+            circlator_logs: path where a summary of log files from circlator will be saved.
+            end_repaired_contigs_filepath: path where the end-repaired contigs will be saved.
+            end_repair_report_filepath: path where the end repair report will be saved.
+            verbose_logfile: path to a logfile where shell script logs will be saved.
+        """
+        self.linking_outdir_base = os.path.join(output_dir, 'contigs')
+        self.circular_contig_tmp_fasta = os.path.join(self.linking_outdir_base, 'circular_input.fasta')
+        self.bam_filepath = os.path.join(output_dir, 'long_read.bam')
+        self.circlator_logs = os.path.join(output_dir, 'circlator_logs')
+        self.end_repaired_contigs_filepath = os.path.join(output_dir, 'repaired.fasta')
+        self.end_repair_report_filepath = os.path.join(output_dir, 'repaired_info.tsv')
+        self.verbose_logfile = os.path.join(output_dir, 'verbose.log')
+
+
+class StitchDirectories:
+    """
+    A class containing directory paths that are used during a single iteration of the assembly and stitch process.
+    """
+
+    def __init__(self, linking_outdir: str, length_threshold: int = None, make_dirs: bool = False):
+        """
+        Instantiate a StitchPaths object.
+
+        :param linking_outdir: output directory for the overall analysis.
+        :param length_threshold: length (bp) that the original contig will be subset to for the given analysis.
+        :param make_dirs: if True, makes the directories whose paths are specified by the object, if the directories
+                          do not already exist.
+        """
+        """
+        Description of constant attributes:
+            linking_outdir: output directory for the overall analysis.
+            log_dir_base: directory where key log files will be saved.
+            length_outdir: output directory for a specific length threshold.
+            log_dir: directory where log files for a specific length threshold will be saved.
+        """
+        self.linking_outdir = linking_outdir
+        self.log_dir_base = os.path.join(linking_outdir, 'logs')
+        if make_dirs:
+            os.makedirs(self.linking_outdir, exist_ok=True)
+            os.makedirs(self.log_dir_base, exist_ok=True)
+
+        if length_threshold:
+            self.set_length_iteration_dirs(length_threshold, make_dirs=make_dirs)
+        else:
+            self.length_outdir = None
```

Comment:
I am pretty sure I am not handling these optional directory paths in an ideal way. Some directory paths I want to store in `StitchDirectories` can only be known in some cases (i.e., if a length threshold is known for a given assembly run; see the `iterate_linking_contig_ends` function for details). I define a setter, `set_length_iteration_dirs`, that can define these optional directories as attributes for a `StitchDirectories` object after the `StitchDirectories` object is instantiated. If the length threshold is not provided when the object is instantiated, then these optional attributes are set to `None` until the setter is run. @LeeBergstrand Do you have any feedback on this approach? For example, would it be better to set the optional attributes dynamically in `StitchDirectories`? Or do you think it would be cleaner for me to make some kind of sub-class from `StitchDirectories` that includes these attributes? Thanks.

Reply:
@LeeBergstrand Nice ideas -- yes, this sounds closer to what I was aiming for. I'll try implementing this soon.

Reply:
Now implemented -- will ask for your review once this round of refactoring is finished. Thanks!

Reply:
@jmtsuji Updated in https://github.com/rotary-genomics/rotary-utils/pull/13 to use Python properties.